addons/website/static/src/components/dialog/seo.js

```diff
@@ -17,8 +21,122 @@ const seoContext = reactive({
     seoName: '',
     metaImage: '',
     defaultTitle: '',
+    updatedAlts: [],
+    brokenLinks: [],
 });
 
+const getSeo = async (self, onlyKeywords = false) => {
+    const pageTextContentEl = self.website.pageDocument.documentElement.querySelector("#wrap");
+    const lang = self.state.language || "en";
+    const tagWeights = {
+        h1: 5,
+        h2: 4,
+        h3: 3,
+        a: 2,
+        p: 1,
+    };
+    const maxNGrams = 2;
+
+    const getKeywordsFromText = (text, weight, wordCounts) => {
+        const segmenter = new Intl.Segmenter(lang, { granularity: "word" });
+        const segmentedText = segmenter.segment(text);
+        const words = [...segmentedText]
+            .filter((s) => s.isWordLike)
+            .map((s) => s.segment.toLowerCase())
+            .filter((word) => !/[0-9]+/.test(word));
+        const singleWordsLength = words.length;
+        for (let nGram = maxNGrams; nGram > 1; nGram--) {
+            for (let i = 0; i <= singleWordsLength - nGram; i++) {
+                if (words[i].length > 4 && words[i + nGram - 1].length > 4) {
+                    words.push(words.slice(i, i + nGram).join(" "));
+                }
+            }
+        }
+        if (words) {
+            words
+                .filter((word) => word.length > 4)
+                .forEach((word) => {
+                    if (!wordCounts[word]) {
+                        wordCounts[word] = 0;
+                    }
+                    wordCounts[word] += weight * (word.length - 3);
+                });
+        }
+    };
+
+    const jaccardSimilarity = (str1, str2) => {
+        const set1 = new Set(str1.replace(/\s+/g, "").split(""));
+        const set2 = new Set(str2.replace(/\s+/g, "").split(""));
+
+        const intersection = new Set([...set1].filter((item) => set2.has(item)));
+        const union = new Set([...set1, ...set2]);
+
+        return intersection.size / union.size;
+    };
+
+    const extractKeywords = () => {
```

Comment:
Question: If I add some custom keywords and then click on the "Auto Fill" button, the added keywords will be removed, is this behavior OK for you ?

Reply:
"Auto fill" is different than "Auto add". What do you think? Maybe @lebl-odoo has an answer to that?

Reply:
IMO it makes sense since we want to fill automatically the SEO parameters. If the user don't agree with what we propose, he still be able to make some changes or cancel changes.

Moreover, since the feature is on top, I don't think people will fill in the fields and then click on auto fill.